Add blog post about building CLI tools with `pkl run` (#118)
apple/pkl-lang.org: blog/modules/ROOT/pages/building-cli-tools-with-pkl.adoc

// end::excerpt[]

CLI tools written in Pkl provide the same basic I/O capabilities that normal Pkl evaluation does: writing to standard output and the filesystem and reading resources, including via xref:main:language-reference:index.adoc#external-readers[external readers], but there are a couple key differences from regular evaluation:

* File output is relative to the command's working directory, not `--multiple-file-output-path`.
* Imports may be specified dynamically in terms of CLI options using link:https://pkl-lang.org/package-docs/pkl/current/Command/Import.html[`Command.Import`].

These capabilities have enabled rewriting the code generation tools for xref:swift:ROOT:codegen.adoc[pkl-swift] and xref:go:ROOT:codegen.adoc[pkl-go].
We now publish them purely as Pkl modules inside their respective packages instead of separately distributed binaries written in each target language that must function cooperatively with the Pkl portion of the code generator.
This reduces the complexity of our code, our release processes, and adoption requirements for downstream users.

== Running Commands

Running CLI tools built with Pkl is simple:

[source,bash]
----

> pkl run <root command module URI> <command options>

Reviewer: ```suggestion
pkl run <module> [command options]
```

----

Every command has automatically generated CLI help via the `--help`/`-h` flag.

Pkl commands distributed inside xref:main:language-reference:index.adoc#project-dependencies[dependencies of a Project] may also be executed directly using the dependency's alias:

[source,bash]
----
pkl run @<dependency alias>/<module> <command options>
----

> On *nix system, commands in local files may also use a https://en.wikipedia.org/wiki/Shebang_(Unix)[shebang] to allow direct script execution:

Reviewer: ```suggestion
On *nix systems, commands in local files may also use a https://en.wikipedia.org/wiki/Shebang_(Unix)[shebang] to allow direct script execution:
```

[source,pkl]
.my-command.pkl
----
#!/usr/bin/env -S pkl run
// ...
----

[source,bash]
----

> ./my-command.pkl

Reviewer: ```suggestion
chmod +x my-command.pkl
./my-command.pkl
```

----

> In this execution mode, a `shell-completion` subcommand can be used to generate autocomplete scripts for the `bash`, `zsh`, and `fish` shells.

Reviewer: [nit] these are names, not code

```suggestion
In this execution mode, a `shell-completion` subcommand can be used to generate autocomplete scripts for the bash, zsh, and fish shells.
```

== Building Commands

Commands are defined declaratively by extending the `pkl:Command` module.
The module's `output` determines the command's behavior:

* `output.bytes` (by default, derived from `output.text` or `output.value`) is produced to standard output.
* `output.files` entries are written to the filesystem. Unlike `pkl eval`, file paths are relative to the working directory, not a specified output path.

> Command options, both named flags and positional arguments, are defined by the properties of the module's `options` property's class.

Reviewer: ```suggestion
Command options, both named flags and positional arguments, are defined by the declared class on the module's `options` property.
```

Any class or module type may be used for `options`, so commands can easily share and inherit options.
When a command is executed, its `options` property is overridden with the actual parsed option values.

Each property of the command's options class (excluding `hidden` and `local` properties) becomes an option of the command.

> A property's name becomes its name on the command line, its doc comment becomes its CLI help text, and its type determines how the raw user-provided is parsed as a Pkl value.

Reviewer: ```suggestion
A property's name becomes its name on the command line, its doc comment becomes its CLI help text, and its type determines how the raw user-provided value is parsed as a Pkl value.
```

> Properties with nullable types or default values are become optional.

Reviewer: ```suggestion
Properties with nullable types or default values become optional.
```

Here's an example command that shows a variety of different options usage:
[source,pkl]
.my-command.pkl
----
extends "pkl:Command"

options: Options

class Options {
  /// Maximum number of tries to attempt operation before giving up.
  `max-tries`: UInt // <1>

  /// Whether to use cache data locally.
  @BooleanFlag
  cache: Boolean = true // <2>

  /// Log verbosity.
  @CountedFlag { shortName = "v" }
  verbose: Int(this <= 3) // <3>

  /// File paths to operate on.
  @Argument { completionCandidates = "paths" }
  path: Listing<String> // <4>

  /// Duration after which operation will be timed out.
  @Flag { convert = module.convertDuration; metavar = "duration" }
  `connection-timeout`: Duration? // <5>
}
----
<1> Flag `--max-tries=<uint>` (no annotation is equivalent to annotating with `@Flag`); `UInt` validates the input is an integer `>= 0`; Flag is required.
<2> Flags `--cache` and `--no-cache` correspond to `true` and `false` values, respectively; Flag is optional and defaults to `true`.
<3> Flag `--verbose`/`-v` may be specified multiple times, each increasing the option's value; Flag is optional and defaults to `0`.

> <4> Argument `<path>` accepts multiple string values; Shell completion suggests file/directory paths as possible values; Argument is optional and defaults to an empty `Listing`.

Reviewer: ```suggestion
<4> Argument `<path>` accepts multiple string values; shell completion suggests file/directory paths as possible values; Argument is optional and defaults to an empty `Listing`.
```

This command's generated CLI help looks like this:
[source,terminaloutput]
----
$ pkl run my-command.pkl
Usage: my-command.pkl [<options>] [<path>]... <command> [<args>]...

Options:
  --max-tries=<uint>    Maximum number of tries to attempt operation before
                        giving up.
  --cache / --no-cache  Whether to use cache data locally.
  -v, --verbose         Log verbosity.
  --connection-timeout=<duration>
                        Duration after which operation will be timed out.
  -h, --help            Show this message and exit

Arguments:
  <path>  File paths to operate on.

Commands:
  shell-completion  Generate a completion script for the given shell
----

=== Subcommands

Commands may also have subcommands.
Subcommands are also Pkl modules that extend `pkl:Command`.

[source,pkl]
.my-command.pkl
----
extends "pkl:Command"

command {
  subcommands {
    import("my-subcommand.pkl")
  }
}

// ...
----

[source,pkl]
.my-subcommand.pkl
----
extends "pkl:Command"

import "my-command.pkl"

parent: `my-command`

// ...
----

Like root commands, when a subcommand is executed, its `options` property is overridden with the actual parsed option values.

> Similarly, the `parent` property is set to the instantiated parent command module with _its_ `options`property set (and `parent`, if applicable).

Reviewer: ```suggestion
Similarly, the `parent` property is set to the instantiated parent command module with _its_ `options` property set (and `parent`, if applicable).
```

Overriding the type of the `parent` property is optional; it asserts that there is a parent command and the parent is a specific command, which can simplify code for complex CLIs.
The `root` property may be overridden similarly.

This subcommand can then be executed:

[source,bash]
----

> pkl run my-command.pkl <root command options> my-subcommand <subcommand options>

Reviewer: ```suggestion
pkl run my-command.pkl [root command options] my-subcommand [subcommand options]
```

----

=== Dynamic Imports

One of the key differentiators between regular Pkl evaluation and CLI commands is that commands offer a form of dynamic importing.
Normal Pkl import statements (`import "<uri>"`) and expressions (`import("<uri>")`) only accept string literals, not arbitrary expressions.
Command options may return link:https://pkl-lang.org/package-docs/pkl/current/Command/Import.html[`Import`] values from option `convert` and `transformAll` functions to trigger dynamic imports.

One example of using dynamic imports is the pkl-swift code generator.

> This command must accept arbitrary `Module` values and analyze them generate Swift code.

Reviewer: ```suggestion
This command must accept arbitrary `Module` values and analyze them to generate Swift code.
```

> Here, link:https://pkl-lang.org/package-docs/pkl/current/Command/Argument.html#convert[`@Argument#convert`] is set to a function that directly converts the raw string value to a directive to import the module by URI:

Reviewer: [nit] `link:` directive is unnecessary when you already have a link

Also, the separator for a property is `.`

```suggestion
Here, https://pkl-lang.org/package-docs/pkl/current/Command/Argument.html#convert[`Argument.convert`] is set to a function that directly converts the raw string value to a directive to import the module by URI:
```

[source,pkl]
----
/// The Pkl modules to generate as Swift.
@Argument {
  convert = (it) -> new Import { uri = it }
  completionCandidates = "paths"
}
pklInputModules: Listing<Module>?
----

=== Advanced Patterns

> These capabilities compose to enable a some patterns that may not be obvious but are extremely useful!

Reviewer: ```suggestion
These capabilities compose to enable some patterns that may not be obvious but are extremely useful!
```

        new Import { uri = "./generator-settings.pkl" }
      else
        new GeneratorSettings {} // <4>
  completionCandidates = "paths"
}
`generator-settings`: GeneratorSettings
----
<1> If a value is supplied, direct Pkl to import it.
<2> If any value is supplied, use it, otherwise fall back to checking the working directory.
<3> If the file exists in the working directory, import it. Note that the absolute `file:` URI is used here because `read` is relative to the enclosing module URI and this command is most frequently executed from inside the `pkl.swift` package.
<4> Otherwise, fall back to an empty/default value.

== Testing Commands

Pkl commands are defined by writing regular modules, which means they can also be tested just like regular Pkl modules!
Test code may import a command module and instantiate it, directly overriding `options` and/or `parent` as needed to mock out user input:

[source,pkl]
----
amends "pkl:test"

import "my-command.pkl"
import "my-subcommand.pkl"

examples {
  ["Test my-command"] {
    (`my-command`) {
      options {
        // Set my-command options here...
      }
    }.output.text
  }
  ["Test my-subcommand"] {

>     (my-subcommand) {

Reviewer: ```suggestion
    (`my-subcommand`) {
```
